lib/em/pure_ruby.rb

```diff
@@ -927,32 +1005,71 @@ def self.connect bind_addr, bind_port, host, port
     def initialize io
       super
       @pending = true
-      @handshake_complete = false
+      @ssl_handshake_state = nil
     end
 
+    TCP_ESTABLISHED = 1 # why isn't this already a const in Socket?
+
     def ready?
-      if RUBY_PLATFORM =~ /linux/
-        io.getsockopt(Socket::SOL_TCP, Socket::TCP_INFO).unpack("i").first == 1 # TCP_ESTABLISHED
+      if defined?(Socket::SOL_TCP) && defined?(Socket::TCP_INFO)
+        # Linux: tcpi_state is an unsigned char
+        #   struct tcp_info {
+        #       __u8    tcpi_state;
+        #       ...
+        #   }
+        sockinfo = io.getsockopt(Socket::SOL_TCP, Socket::TCP_INFO)
+        sockinfo.unpack("C").first == TCP_ESTABLISHED
+      elsif defined?(Socket::IPPROTO_TCP) && defined?(Socket::TCP_CONNECTION_INFO)
+        # NOTE: the following doesn't seem to work (according to GH actions)
+        #
+        # MacOS: tcpi_state is an unsigned char
+        #   struct tcp_connection_info {
+        #       u_int8_t   tcpi_state;     /* connection state */
+        #       ...
+        #   }
+        sockinfo = io.getsockopt(Socket::IPPROTO_TCP, Socket::TCP_CONNECTION_INFO)
+        sockinfo.unpack("C").first == TCP_ESTABLISHED
       else
-        io.getsockopt(Socket::SOL_SOCKET, Socket::SO_ERROR).unpack("i").first == 0 # NO ERROR
+        sockerr = io.getsockopt(Socket::SOL_SOCKET, Socket::SO_ERROR)
+        sockerr.unpack("i").first == 0 # NO ERROR
       end
     end
 
-    def handshake_complete?
-      if !@handshake_complete && io.respond_to?(:state)
-        if io.state =~ /^SSLOK/
-          @handshake_complete = true
-          EventMachine::event_callback uuid, SslHandshakeCompleted, ""
-          EventMachine::event_callback uuid, SslVerify, io.peer_cert.to_pem if io.peer_cert
-        end
-      else
-        @handshake_complete = true
-      end
-      @handshake_complete
+    def eventable_read
+      check_handshake_complete and super
+    end
+
+    def eventable_write
+      check_handshake_complete and super
+    end
+
+    def start_tls(ssl_io)
+      self.io = ssl_io
+      @ssl_handshake_state = :init
+      check_handshake_complete
+    end
+
+    def check_handshake_complete
+      return true if ssl_handshake_state.nil? || ssl_handshake_state == :done
+      is_server ? io.accept_nonblock : io.connect_nonblock
+      @ssl_handshake_state = :done
+      EventMachine::event_callback uuid, SslHandshakeCompleted, ""
+      true
+    rescue SSLConnectionWaitReadable
+      @ssl_handshake_state = :wait_readable
+      false
+    rescue SSLConnectionWaitWritable
+      @ssl_handshake_state = :wait_writable
+      false
+    rescue OpenSSL::SSL::SSLError => error
+      warn "SSL Error in EventMachine check_handshake_complete: #{error}"
+      @ssl_handshake_state = error
+    rescue => error
+      warn "#{error.class} in EventMachine check_handshake_complete: #{error}"
+      @ssl_handshake_state = error
     end
 
     def pending?
-      handshake_complete?
```

Comment:
Seems likely this method was internal use only, but calling it out to think/confirm

Reply:
Yeah, `#handshake_complete?` seems to have only existed in the pure_ruby selector... it seemed similar in function to `ConnectionDescriptor::_CheckHandshakeStatus` in the cpp.  I renamed it to (or replaced it with) `#check_handshake_complete`, because calling predicate methods for their side effects confuses me (also it matches the name of its related cpp method a little closer now).